File: app/src/main/java/github/gggxbbb/tujian_dev/java/PicsAdapter.java
package github.gggxbbb.tujian_dev.java;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import github.gggxbbb.tujian_dev.DetailActivity;
import github.gggxbbb.tujian_dev.R;
import github.gggxbbb.tujian_dev.tools.TujianPic;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

public class PicsAdapter extends RecyclerView.Adapter<PicsAdapter.BeautyViewHolder> {

    private Context mContext;

    private List<TujianPic> data;

    public PicsAdapter(List<TujianPic> data, Context context) {
        this.data = data;
        this.mContext = context;
    }

    @NotNull
    @Override
    public BeautyViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.activity_main_item, parent, false);
        return new BeautyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NotNull final BeautyViewHolder holder, int position) {
        final TujianPic beauty = data.get(position);
        Glide.with(mContext).load(beauty.getLink()).listener(new RequestListener<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                holder.onLoading.setVisibility(View.GONE);
                holder.info_text.setText(Objects.requireNonNull(e).getMessage());
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                return false;
            }
        }).into(holder.pic);
        holder.pic.setMinimumHeight(beauty.getHeight() * (holder.pic.getWidth() / beauty.getWidth()));
        holder.content.setText(beauty.getContent());
        holder.title.setText(beauty.getTitle());
        holder.info.setText(String.format("%s %s %s×%s @ %s", beauty.getDate(), beauty.getTNAME(mContext), beauty.getWidth(), beauty.getHeight(), beauty.getUsername()));
        holder.linearLayout.setBackgroundColor(Color.parseColor(beauty.getThemeColor()));
        holder.content.setTextColor(Color.parseColor(beauty.getTextColor()));
        holder.content.setLinkTextColor(Color.parseColor(beauty.getTextColor()));
        holder.title.setTextColor(Color.parseColor(beauty.getTextColor()));
        holder.info.setTextColor(Color.parseColor(beauty.getTextColor()));
        holder.root.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, DetailActivity.class);
                intent.putExtra("pic", beauty.getString());
                mContext.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    static class BeautyViewHolder extends RecyclerView.ViewHolder {
        ImageView pic;
        TextView title, info, content,info_text;
        LinearLayout linearLayout;
        View root;
        ProgressBar onLoading;

        BeautyViewHolder(View itemView) {
            super(itemView);
            pic = itemView.findViewById(R.id.pic);
            title = itemView.findViewById(R.id.pic_title);
            info = itemView.findViewById(R.id.pic_info);
            content = itemView.findViewById(R.id.pic_content);
            linearLayout = itemView.findViewById(R.id.info);
            root = itemView.getRootView();
            onLoading = itemView.findViewById(R.id.onLoading);
            info_text = itemView.findViewById(R.id.info_text);
        }
    }

    public void addItem(TujianPic tujianPic) {
        data.add(tujianPic);
        notifyItemInserted(data.size());
    }

}

File: app/src/main/java/github/gggxbbb/tujian_dev/tools/PicsAdapter.kt
package github.gggxbbb.tujian_dev.tools

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import coil.load
//import com.zzhoujay.richtext.RichText
import github.gggxbbb.tujian_dev.DetailActivity
import github.gggxbbb.tujian_dev.R
import github.gggxbbb.tujian_dev.tools.PicsAdapter.TujianViewHolder

class PicsAdapter(
    private val data: MutableList<TujianPic>,
    private val activity: Activity
) : RecyclerView.Adapter<TujianViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TujianViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.activity_main_item, parent, false)
        return TujianViewHolder(view)
    }

    override fun onBindViewHolder(holder: TujianViewHolder, position: Int) {
        val tujianPic = data[position]
        holder.pic.load(tujianPic.getLink()) {
            crossfade(true)
            listener(onStart = {
                holder.onLoading.visibility = View.VISIBLE
            },
                onSuccess = { _, _ ->
                    holder.onLoading.visibility = View.GONE
                },
                onError = { _, e ->
                    holder.onLoading.visibility = View.GONE
                    holder.infoText.text = e.message
                })
        }
        holder.pic.minimumHeight = tujianPic.getHeight() * (holder.pic.width / tujianPic.getWidth())
        //RichText.fromMarkdown(tujianPic.getContent()).into(holder.content)
        //holder.content.text = tujianPic.getContent()
        holder.title.text = tujianPic.getTitle()
        /*holder.info.text = String.format(
            "%s %s %s×%s @ %s",
            tujianPic.getDate(),
            tujianPic.getTNAME(),
            tujianPic.getWidth(),
            tujianPic.getHeight(),
            tujianPic.getUsername()
        )*/
        holder.linearLayout.setBackgroundColor(Color.parseColor(tujianPic.getThemeColor()))
        /*holder.content.setTextColor(Color.parseColor(tujianPic.getTextColor()))
        holder.content.setLinkTextColor(Color.parseColor(tujianPic.getTextColor()))*/
        holder.title.setTextColor(Color.parseColor(tujianPic.getTextColor()))
        /*holder.info.setTextColor(Color.parseColor(tujianPic.getTextColor()))*/
        holder.root.setOnClickListener {
            val intent = Intent(activity, DetailActivity::class.java)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                intent.addFlags(Intent.FLAG_ACTIVITY_LAUNCH_ADJACENT)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                intent.addFlags(Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
            }
            intent.putExtra("pic", tujianPic.getString())
            activity.startActivity(intent)
        }

        stateRestorationPolicy = StateRestorationPolicy.PREVENT_WHEN_EMPTY

    }

    override fun getItemCount(): Int {
        return data.size
    }

    class TujianViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var pic: ImageView = itemView.findViewById(R.id.pic)
        var title: TextView = itemView.findViewById(R.id.pic_title)
        var info: TextView = itemView.findViewById(R.id.pic_info)
        var content: TextView = itemView.findViewById(R.id.pic_content)
        var infoText: TextView = itemView.findViewById(R.id.info_text)
        var linearLayout: LinearLayout = itemView.findViewById(R.id.info)
        var root: View = itemView.rootView
        var onLoading: ProgressBar = itemView.findViewById(R.id.onLoading)

    }

    fun addItem(tujianPic: TujianPic) {
        data.add(tujianPic)
        notifyItemInserted(data.size)
    }

}